### ISSUE TITLE
SSML prosodmostly unsupported?

### ISSUE THREAD

**Issue by navid-zamani** [espeak-ng/espeak-ng#1711]
I’ve found much of SSML to not do anything in espeak-ng 1.50.  
(Unfortunately Linux Mint does not offer a newer version yet, but I hope you can reproduce it anyway.-

E.g. in the following, the pitch is the same for both sentences:

    espeak-ng  -m '<prosody pitch="x-low">Low pitch.</prosody> <prosody pitch="x-high">High pitch.</prosody>' 

Range and rate do not work either.

**Comment by Myuqou** [comment]
I've been running into the same problem on both Windows and Linux.  The documentation for building eSpeak-NG from source states:

-------------------------------------------------------------------------
Optionally, you need:
    sox (http://sox.sourceforge.net/) to enable audio output for SSML tag
-------------------------------------------------------------------------

...but it doesn't go into any specifics on what to do with SoX, so I have no idea if I have everything set up correctly.

On Windows, I dumped the SoX source code into the eSpeak-NG\src folder before compiling and on that version I was able to get the prosody pitch tags to work half the time, if I specify an actual number instead of "x-low" or "x-high".

On Linux I just tried installing the already compiled versions of SoX and eSpeak-NG using apt, but that didn't work at all.

**Comment by navid-zamani** [comment]
@Myuqou:

Yes, you are right, it does work with numeric values. With a self-build espeak-ng, the espeak-ng in my system’s package manager, and even with the old espeak in my system’s package manager!

**The strange thing is, that it *is* parsed, and turned into numeric values:**

https://github.com/espeak-ng/espeak-ng/blob/master/src/libespeak-ng/ssml.c

Pretty much all of SSML is parsed into speech parameters just fine in there.

I’m not that great at reading plain old C. But it looks like a bug to me. (That I already spent way too much time on, as a mere user.)

Also, when I’m building it manually, it seems to not be able to output sound by itself *at all*. Only with `--stdout` does it work. (And the espeak-ng in my system’s package manage, while being able to output sound by itself, fails to handle sample rate conversion, resulting in lost of ugly noise.)

---

#### On SoX:

I guess for Linux, this is where it handles building and the configuration options it accepts:  
https://github.com/espeak-ng/espeak-ng/blob/master/configure.ac

And here you can see how it is actually built and installed, using github’s “workflows” functionality:  
https://github.com/espeak-ng/espeak-ng/blob/master/.github/workflows/autoconf.yml

**I see no mention of SoX in either.** But several other things it can use for output. (See the bottom of the first file.)

---

#### Note on what to do with dependencies, in general:

The usual method when software says it needs other software, is not mentioned because on anything Unixy, it is obvious: It means it is a dependency, as in: It would be listed in the dependencies for the package in the package manager.  
(Note that the build-time dependencies and run-time dependencies can and usually do differ.)
So yes, if you do it manually, you install that software, using the normal system package manager. So the files go into the normal system locations, and build systems know where to look for them.  

**BUT: If you want to build something, you usually need to install the *source* package (too).** Usually, that has a `-dev` at the end of the name, or the like. Except for source distributions that compile everything in any case, where you just install the normal (library) package,

You can also pass custom dependency/library locations to the `./configure` script, normally.

**On Windows,** package managers (esp. with dependencies) haven’t really been invented yet, and software is usually distributed in a monolithic way, with everything the software needs in its own directory. While runtime libraries *can* go to the system library directories if you want them to be shared with other software to save resources.  
But the build system (in this case Visual Studio) should allow similar custom location settings. I don’t know what interfaces VS has, but there will definitely be configuration for optional dependencies in there. There you can see the path(s) it looks in.

The easiest way is often to just quickly read the build file with a text editor. If you know the assumptions of the particular build system.

**Comment by jaacoppi** [comment]
SSML prosody implementation is known to be buggy. The root cause is sentence termination detection. There are many bug reports about it already.

sox isn't used anymore. The documentation needs to be updated.

**Comment by navid-zamani** [comment]
Hey @jaacoppi: Thanks for the feedback. What does ”sentence termination” mean in this context?

I only output short (status) lines. Single sentences. So it’s probably easy for me to work around it, and I would like to be able to do that so I can finish the bug in my script.

Also, I did of course search or similar bugs, as I know the annoyance of duplicates well, but did find none. So I’m sorry to have caused a duplicate, but apparently have to idea what to search for. Can you point me to what you consider the core bug for this, so I can follow it instead and we can close this?

**Comment by jaacoppi** [comment]
I don't know about core bug, but at least #410 , #737 and #1533 appear with search for "ssml".

To simplify, sentence termination is done after punctuation like . , ! ? and so on. Technically, ReadClause() reads the input string until it finds a sentence terminator. SSML parsing is done inside ReadClause. In some cases when an SSML input contains a sentence terminator the internal state gets mixed up  and the tags aren't processed. You can easily test this by comparing:
`espeak-ng  -m <prosody pitch="x-low">Low pitch.</prosody> <prosody pitch="x-high">High pitch.</prosody>`
to
`espeak-ng  -m '<prosody pitch="x-low">Low pitch </prosody> <prosody pitch="x-high">High pitch</prosody>`
